src/Orleans.Core.Abstractions/IDs/SpanId.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Serialization;
+using System.Text;
+using Orleans.Concurrency;
+
+namespace Orleans.Runtime
+{
+    [Immutable]
+    [Serializable]
+    [StructLayout(LayoutKind.Auto)]
+    public readonly struct SpanId : IEquatable<SpanId>, IComparable<SpanId>, ISerializable
+    {
+        private readonly byte[] _value;
+        private readonly int _hashCode;
+
+        public SpanId(byte[] value)
+        {
+            _value = value;
+            _hashCode = GetHashCode(value);
+        }
+
+        public SpanId(byte[] value, int hashCode)
```

Comment:
The idea is that these will be used for serialization. I will make the constructor private and add an UnsafeCreate method instead.

Reply:
Cool. Yeah I totally get that serialization needs this for best performance. I was mostly worried about this being a potential foot gun since there is nothing in the signature to indicate it is potentially risky to use if you don't understand the implementation.